docs/source/examples/checkpointer.rst

```diff
@@ -0,0 +1,451 @@
+
+
+For this checkpointer to work correctly, we assume that checkpoint_dir contains the necessary checkpoint
+and json files. The easiest way to make sure everything works correctly is to use the following flow:
+
+- Download the model from the HF repo using tune download. By default, this will ignore the "safetensors"
+  files.
+
+    |
+
+    .. code-block:: bash
+
+        tune download meta-llama/Llama-2-7b
+        --output-dir <checkpoint_dir>
+        --hf-token <hf-token>
+
+- Use ``output_dir`` above as the ``checkpoint_dir`` for the checkpointer.
+
+|
+
+The following snippet explains how the MetaCheckpointer is setup in TorchTune config files.
+
+.. code-block:: yaml
+
+    checkpointer:
+
+        # checkpointer to use
+        _component_: torchtune.utils.FullModelMetaCheckpointer
+
+        # directory with the checkpoint files
+        # this should match the output_dir above
+        checkpoint_dir: <checkpoint_dir>
+
+        # checkpoint files. For the llama2-7b model we have
+        # a single .pth file
+        checkpoint_files: [consolidated.00.pth]
+
+        # if we're restarting a previous run, we need to specify
+        # the file with the checkpoint state. More on this in the
+        # next section
+        recipe_checkpoint: null
+
+        # dir for saving the output checkpoints. Usually set
+        # to be the same as checkpoint_dir
+        output_dir: <checkpoint_dir>
+
+        # model_type which specifies how to convert the state_dict
+        # into a format which TorchTune understands
+        model_type: LLAMA2
+
+    # set to True if restarting training
+    resume_from_checkpoint: False
+
+|
+
+**TorchTuneCheckpointer**
+
+This checkpointer reads and writes checkpoints in a format that is compatible with TorchTune's
+model definition. This does not perform any state_dict conversions and is currently used either
+for testing or for loading quantized models for generation.
+
+|
+
+
+Intermediate vs Final Checkpoints
+---------------------------------
+
+TorchTune Checkpointers support two checkpointing scenarios:
+
+**End-of-training Checkpointing**
+
+The model weights at the end of a completed training
+run are written out to file. The checkpointer ensures that the output checkpoint
+files have the same keys as the input checkpoint file used to begin training. The
+checkpointer also ensures that the keys are paritioned across the same number of
+files as the original checkpoint. The output state dict has the following
+standard format:
+
+  .. code-block:: python
+
+            {
+                "key_1": weight_1,
+                "key_2": weight_2,
+                ...
+            }
+
+
+**Mid-training Chekpointing**.
+
+If checkpointing in the middle of training, the output checkpoint needs to store additional
+information to ensure that subsequent training runs can be correctly restarted. In addition to
+the model checkpoint files, we output a ``recipe_state.pt`` file for intermediate
```

Comment:
Its a good use case, but seems like an edge case for now? Happy to revisit this as well based on user info. The problem is that recipe_state can be quite a big cpt since we capture opt state as well. So we end up with ~13.5GB checkpoint/epoch for weights and then ~27GB checkpoint/epoch for opt state. This can be quite a lot esp when users dont have much disk space